src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        dataArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index is invalid");
+        }
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        if (index == size) {
+            dataArray[size++] = value;
+        } else {
+            System.arraycopy(dataArray, index, dataArray, index + 1, size++);
+            dataArray[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int oldSize = size;
+        int neededSize = size + list.size();
+        int indexOfValueFromList = START_INDEX_VALUE;
+        if (size == dataArray.length || dataArray.length < neededSize) {
+            dataArray = growToNeededSize(neededSize);
+        }
+        for (int i = oldSize; i < neededSize; i++) {
+            dataArray[i] = list.get(indexOfValueFromList++);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) dataArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        dataArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return remove(dataArray[index]);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T[] temp = (T[]) new Object[Math.max(DEFAULT_CAPACITY, dataArray.length - 1)];
+        for (int i = 0; i < dataArray.length; i++) {
+            if (element == null && dataArray[i] == null
+                    || element != null && element.equals(dataArray[i])) {
+                for (int index = 0; index < i; index++) {
+                    temp[index] = dataArray[index];
+                }
+                for (int j = i; j < dataArray.length - 1; j++) {
+                    temp[j] = dataArray[j + 1];
+                }
+                System.arraycopy(temp, 0, dataArray, 0, temp.length);
+                size--;
+                return element;
+            }
+        }
+        throw new NoSuchElementException("This element does not belong to this list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T[] grow() {
+        double newCapacity = dataArray.length * GROW_INDEX;
+        T[] newArray = (T[]) new Object[(int) newCapacity];
+        System.arraycopy(dataArray, 0, newArray, 0, dataArray.length);
+        return newArray;
+    }
+
+    private T[] growToNeededSize(int neededSize) {
```

Comment:
The `growToNeededSize` method is unnecessary. You should resize the array in a separate method, but this logic should be a part of the `grow` method itself, which ensures that the array has enough capacity.